{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cached data available...\n",
      "[('hsa-miR-1-3p', 983), ('hsa-miR-22-3p', 659), ('hsa-miR-221-3p', 532), ('hsa-miR-132-3p', 514), ('hsa-miR-212-5p', 447), ('hsa-miR-150-5p', 369), ('hsa-miR-342-3p', 326), ('hsa-miR-409-3p', 325), ('hsa-miR-486-5p', 178), ('hsa-miR-210-3p', 48)]\n",
      "(4400, 17)\n",
      "Group  stable_hsp-hit\n",
      "Group size  280 (1697)\n",
      "Target hits in group  391\n",
      "Avg. target hits per gene  1.4\n",
      "====\n",
      "Group  deg_hsp-hit\n",
      "Group size  79 (438)\n",
      "Target hits in group  119\n",
      "Avg. target hits per gene  1.51\n",
      "====\n",
      "Group  deg_no-hit\n",
      "Group size  340 (1971)\n",
      "Target hits in group  482\n",
      "Avg. target hits per gene  1.42\n",
      "====\n",
      "Group  stable_no-hit\n",
      "Group size  2391 (15985)\n",
      "Target hits in group  3408\n",
      "Avg. target hits per gene  1.43\n",
      "====\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import os as os\n",
    "import collections as col\n",
    "import itertools as itt\n",
    "import pickle as pck\n",
    "import json as js\n",
    "import time as ti\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.lines as mlines\n",
    "\n",
    "import numpy as np\n",
    "import numpy.random as rng\n",
    "import scipy.stats as stats\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# What does this do?\n",
    "# Plot boxplots of minimum TPM\n",
    "# log2 fold-change for genes\n",
    "# stratified by HSP hits\n",
    "# (body/enhancer)\n",
    "\n",
    "date = '20180709'\n",
    "\n",
    "run_plot_gene_enh_hsp = False\n",
    "run_dump_genetrail_lists = True\n",
    "\n",
    "save_figures = True\n",
    "\n",
    "journal = 'ox_bioinf'\n",
    "res = 'print_lo'\n",
    "fig_sizes = js.load(open('/home/pebert/work/code/mpggit/statediff/annotation/misc/fig_sizes.json'))\n",
    "fig_sizes = fig_sizes[journal]\n",
    "resolution = fig_sizes['resolution']\n",
    "\n",
    "segmentations = ['cmm18', 'esc18', 'ecs10']\n",
    "scorings = ['emission', 'replicate']\n",
    "\n",
    "sns.set(style='white',\n",
    "        font_scale=1.5,\n",
    "        rc={'font.family': ['sans-serif'],\n",
    "            'font.sans-serif': ['DejaVu Sans']})\n",
    "\n",
    "fhgfs_base = '/TL/deep/fhgfs/projects/pebert/thesis/projects/statediff'\n",
    "cache_dir = os.path.join(fhgfs_base, 'caching/notebooks')\n",
    "\n",
    "isect_folder = os.path.join(fhgfs_base, 'bedtools/deep/inv_isect')\n",
    "de_gene_folder = os.path.join(fhgfs_base, 'deseq/deep')\n",
    "bed_gene_folder = os.path.join(fhgfs_base, 'deseq/bed_out')\n",
    "\n",
    "genetrail_out = os.path.join(fhgfs_base, 'genetrail/input_lists', date)\n",
    "\n",
    "tpm_file = os.path.join(fhgfs_base, 'salmon', 'deep', 'agg_gene_tpm.h5')\n",
    "\n",
    "base_out = '/TL/deep-external01/nobackup/pebert/cloudshare/mpiinf/phd/chapter_projects/statediff'\n",
    "\n",
    "fig_supp = os.path.join(base_out, 'figures', 'pub', 'supp')\n",
    "fig_main = os.path.join(base_out, 'figures', 'pub', 'main')\n",
    "fig_collect = os.path.join(base_out, 'figures', 'pub', 'collection')\n",
    "                   \n",
    "\n",
    "def load_gene_table(fpath, genetype):\n",
    "    \n",
    "    genes = dict()\n",
    "    with open(fpath, 'r') as table:\n",
    "        _ = table.readline()\n",
    "        for line in table:\n",
    "            cols = line.strip().split('\\t')\n",
    "            try:\n",
    "                name, symbol = cols[3].split('.')[0], cols[6]\n",
    "            except IndexError:\n",
    "                break\n",
    "            log2fc = cols[4]\n",
    "            try:\n",
    "                gene = genes[name]\n",
    "            except KeyError:\n",
    "                gene = {'name': name, 'symbol': symbol,\n",
    "                        'log2fc': float(log2fc), 'genetype': genetype,\n",
    "                        'body_hsp': 0}\n",
    "                genes[name] = gene\n",
    "            if int(cols[-1]) > 0:\n",
    "                gene['body_hsp'] += 1\n",
    "    df = pd.DataFrame.from_records(list(genes.values()), index='name')\n",
    "    return df\n",
    "\n",
    "\n",
    "def load_enhancer_table(fpath):\n",
    "    \n",
    "    genes = dict()\n",
    "    with open(fpath, 'r') as table:\n",
    "        _ = table.readline()\n",
    "        for line in table:\n",
    "            cols = line.strip().split('\\t')\n",
    "            try:\n",
    "                enh_id, ens_id, symbol, intra = cols[3], cols[6], cols[7], int(cols[10])\n",
    "            except IndexError:\n",
    "                break\n",
    "            try:\n",
    "                gene = genes[ens_id]\n",
    "            except KeyError:\n",
    "                gene = {'name': ens_id, 'enhancer': set(), 'enh_hits': set()}\n",
    "                genes[ens_id] = gene\n",
    "            gene['enhancer'].add((enh_id, intra))\n",
    "            if int(cols[-1]) > 0:\n",
    "                gene['enh_hits'].add((enh_id, intra))\n",
    "                \n",
    "    for k, v in genes.items():\n",
    "        v['total_enh'] = len(v['enhancer'])\n",
    "        v['intra_enh'] = sum([t[1] for t in v['enhancer']])\n",
    "        v['inter_enh'] = v['total_enh'] - v['intra_enh']\n",
    "        v['total_hits'] = len(v['enh_hits'])\n",
    "        v['intra_hits'] = len([t for t in v['enh_hits'] if t[1] == 1])\n",
    "        v['inter_hits'] = v['total_hits'] - v['intra_hits']\n",
    "        del v['enhancer']\n",
    "        del v['enh_hits']\n",
    "        \n",
    "    df = pd.DataFrame.from_records(list(genes.values()), index='name')\n",
    "    return df\n",
    "    \n",
    "                \n",
    "def load_expression_data(s1, s2):\n",
    "    \n",
    "    with pd.HDFStore(tpm_file, 'r') as hdf:\n",
    "        load_keys = [k for k in hdf.keys() if k.endswith(s1) or k.endswith(s2)]\n",
    "        assert len(load_keys) == 2, 'data missing'\n",
    "        data = pd.concat([hdf[load_keys[0]], hdf[load_keys[1]]], axis=1, ignore_index=False)\n",
    "        data['min_tpm'] = data.min(axis=1)\n",
    "        data = data.loc[:, ['min_tpm']].copy()\n",
    "    return data\n",
    "        \n",
    "    \n",
    "def cache_overlap_data(rootfolder, cache_file):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    \n",
    "    defiles = [f for f in os.listdir(rootfolder) if f.startswith('deep_degenes_') and f.endswith('.tsv')]\n",
    "    defiles = sorted(defiles)\n",
    "    \n",
    "    for defile in defiles:\n",
    "        depath = os.path.join(rootfolder, defile)\n",
    "        de_genes = load_gene_table(depath, 'diff')\n",
    "        stpath = depath.replace('_degenes_', '_stgenes_')\n",
    "        st_genes = load_gene_table(stpath, 'stable')\n",
    "        all_genes = pd.concat([de_genes, st_genes], axis=0, ignore_index=False)\n",
    "        enhpath = depath.replace('_degenes_', '_enh_')\n",
    "        enh_genes = load_enhancer_table(enhpath)\n",
    "        parts = defile.split('.')[0].split('_')\n",
    "        seg, comp, scoring = parts[4], parts[5] + '_vs_' + parts[7], parts[8]\n",
    "        \n",
    "        tpm_data = load_expression_data(parts[5], parts[7])\n",
    "        tpm_data = tpm_data.loc[tpm_data.index.isin(all_genes.index), :]\n",
    "        \n",
    "        joined = pd.concat([all_genes, enh_genes, tpm_data], axis=1,\n",
    "                           ignore_index=False, sort=True)\n",
    "        joined.fillna(0, inplace=True)\n",
    "        \n",
    "        store_path = os.path.join(seg, scoring, comp)\n",
    "        with pd.HDFStore(cache_file, 'a') as hdf:\n",
    "            hdf.put(store_path, joined, format='table')\n",
    "            \n",
    "    return cache_file\n",
    "    \n",
    "    \n",
    "def prepare_plot_data(data, data_type):\n",
    "    \n",
    "    boxes = []\n",
    "    box_labels = []\n",
    "    # genes w/o body hit and no enhancer hit\n",
    "    select = np.logical_and(data['body_hsp'] == 0, data['total_hits'] == 0)\n",
    "    sub = data.loc[select, data_type].abs()\n",
    "    boxes.append(sub)\n",
    "    box_labels.append('B:0  E:0 ({})'.format(sub.size))\n",
    "    \n",
    "    # genes w/o body hit AND at least 1 enhancer hit\n",
    "    select = np.logical_and(data['body_hsp'] == 0, data['total_hits'] == 1)\n",
    "    sub = data.loc[select, data_type].abs()\n",
    "    boxes.append(sub)\n",
    "    box_labels.append('B:0  E:1 ({})'.format(sub.size))\n",
    "    \n",
    "    select = np.logical_and(data['body_hsp'] == 0, data['total_hits'] == 2)\n",
    "    sub = data.loc[select, data_type].abs()\n",
    "    boxes.append(sub)\n",
    "    box_labels.append('B:0  E:2 ({})'.format(sub.size))\n",
    "    \n",
    "    select = np.logical_and(data['body_hsp'] == 0, data['total_hits'] >= 3)\n",
    "    sub = data.loc[select, data_type].abs()\n",
    "    boxes.append(sub)\n",
    "    box_labels.append('B:0  E:3+ ({})'.format(sub.size))\n",
    "    \n",
    "    # genes with at least one body hit\n",
    "    select = np.logical_and(data['body_hsp'] > 0, data['total_hits'] == 0)\n",
    "    sub = data.loc[select, data_type].abs()\n",
    "    boxes.append(sub)\n",
    "    box_labels.append('B:1+  E:0 ({})'.format(sub.size))\n",
    "    \n",
    "    # genes with at least one body hit\n",
    "    select = np.logical_and(data['body_hsp'] > 0, data['total_hits'] == 1)\n",
    "    sub = data.loc[select, data_type].abs()\n",
    "    boxes.append(sub)\n",
    "    box_labels.append('B:1+  E:1 ({})'.format(sub.size))\n",
    "    \n",
    "    # genes with at least one body hit\n",
    "    select = np.logical_and(data['body_hsp'] > 0, data['total_hits'] == 2)\n",
    "    sub = data.loc[select, data_type].abs()\n",
    "    boxes.append(sub)\n",
    "    box_labels.append('B:1+  E:2 ({})'.format(sub.size))\n",
    "    \n",
    "    # genes with at least one body hit\n",
    "    select = np.logical_and(data['body_hsp'] > 0, data['total_hits'] >= 3)\n",
    "    sub = data.loc[select, data_type].abs()\n",
    "    boxes.append(sub)\n",
    "    box_labels.append('B:1+  E:3+ ({})'.format(sub.size))\n",
    "    \n",
    "    return boxes, box_labels\n",
    "    \n",
    "    \n",
    "def generate_bins(dataset, stepsize):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    fc_min = dataset['log2fc'].min()\n",
    "    assert fc_min < 0, 'That is unexpected...'\n",
    "    fc_lower = (fc_min // stepsize) * stepsize - stepsize\n",
    "    fc_max = dataset['log2fc'].max()\n",
    "    fc_upper = (fc_max // stepsize) * stepsize + stepsize\n",
    "    bins = np.arange(fc_lower, fc_upper + 0.1, stepsize)\n",
    "    bin_num = [i for i, b in enumerate(bins, start=1)]\n",
    "    return bins, bin_num\n",
    "    \n",
    "\n",
    "def compute_densities(dataset, bins, hit_type):\n",
    "    \n",
    "    histograms = []\n",
    "    labels = []\n",
    "    lo_bounds = [0, 2, 5, 8, 11]\n",
    "    hi_bounds = [0, 3, 6, 9, 1000]\n",
    "    # no overlap\n",
    "    for lo, hi in zip(lo_bounds, hi_bounds):\n",
    "        select = np.logical_and(dataset[hit_type] >= lo, dataset[hit_type] <= hi)\n",
    "        group = dataset.loc[select, 'log2fc']\n",
    "        if group.size == 0:\n",
    "            hist = np.zeros_like(histograms[0])\n",
    "            histograms.append(hist)\n",
    "        else:\n",
    "            hist, bin_edges = np.histogram(group, bins, density=True)\n",
    "            histograms.append(np.concatenate([[0], hist]))\n",
    "        if lo == 0:\n",
    "            labels.append('0: {}'.format(group.size))\n",
    "        elif lo == max(lo_bounds):\n",
    "            labels.append('{}+: {}'.format(lo, group.size))\n",
    "        else:\n",
    "            labels.append('{}-{}: {}'.format(lo, hi, group.size))\n",
    "    \n",
    "    return histograms, labels\n",
    "    \n",
    "    \n",
    "    \n",
    "def create_cdf_plot(datapoints, comp, fkey):\n",
    "    \n",
    "    fig, axes = plt.subplots(figsize=fig_sizes['two_col']['square'],\n",
    "                             nrows=3, ncols=2,\n",
    "                             sharex=False, sharey=False)\n",
    "    stepsize = 0.5\n",
    "    bins, bin_nums = generate_bins(datapoints, stepsize)\n",
    "    \n",
    "    plt.subplots_adjust(wspace=0.05, hspace=0.05)\n",
    "    \n",
    "    data_colors = ['dimgrey', 'dodgerblue', 'limegreen', 'orange', 'magenta']\n",
    "    \n",
    "    # for all enhancers\n",
    "    total_hist, total_labels = compute_densities(datapoints, bins, 'total_hits')\n",
    "    # for intragenic\n",
    "    intra_hist, intra_labels = compute_densities(datapoints, bins, 'intra_hits')\n",
    "    # for intragenic\n",
    "    inter_hist, inter_labels = compute_densities(datapoints, bins, 'inter_hits')\n",
    "    row_title = {0: 'anywhere', 1: 'intragenic', 2: 'intergenic'}\n",
    "    \n",
    "    s1, s2 = comp.split('_vs_')\n",
    "    for row_idx, (data, lab) in enumerate([(total_hist, total_labels),\n",
    "                                           (intra_hist, intra_labels),\n",
    "                                           (inter_hist, inter_labels)]):\n",
    "        ax = axes[row_idx, 0]\n",
    "        if row_idx == 0:\n",
    "            tt = ax.set_title('HSP gene enhancer hits: {} v {}'.format(s1, s2), fontsize=14)\n",
    "            tt.set_position((1, 1))\n",
    "        \n",
    "        leg_handles = []\n",
    "        for c, hist, l in zip(data_colors, data, lab):\n",
    "            if c == 'dimgrey':\n",
    "                style = 'dashed'\n",
    "                zorder = 0\n",
    "            else:\n",
    "                style = 'solid'\n",
    "                zorder = 2\n",
    "            assert np.isclose(hist.sum() * stepsize, 1., atol=1e-6), 'CDF not 1'\n",
    "            ax.plot(bins, np.cumsum(hist) * stepsize,\n",
    "                    lw=2, ls=style, c=c, zorder=zorder,\n",
    "                    label=l)\n",
    "#             ax.axhline(0.75, 0, 1, c='lightgrey',\n",
    "#                        ls='dashed', zorder=0, lw=1)\n",
    "#             ax.axhline(0.5, 0, 1, c='lightgrey',\n",
    "#                        ls='dashed', zorder=0, lw=1)\n",
    "#             ax.axhline(0.25, 0, 1, c='lightgrey',\n",
    "#                        ls='dashed', zorder=0, lw=1)\n",
    "            \n",
    "            ax.axvline(-3, 0, 1, c='black', alpha=0.5,\n",
    "                       ls='dashed', zorder=0, lw=1)\n",
    "            ax.axvline(3, 0, 1, c='black', alpha=0.5,\n",
    "                       ls='dashed', zorder=0, lw=1)\n",
    "            leg_patch = mlines.Line2D([], [], marker='s', markersize=7,\n",
    "                                      lw=0, color=c, label=l)\n",
    "            leg_handles.append(leg_patch)\n",
    "        \n",
    "        if comp == 'HG_vs_Mo':\n",
    "            ax_leg = ax.legend(loc='right', handles=leg_handles,\n",
    "                               fontsize=10, bbox_to_anchor=(1.05, 0.4))\n",
    "        else:\n",
    "            ax_leg = ax.legend(loc='best', handles=leg_handles,\n",
    "                               fontsize=10)\n",
    "        tt = ax_leg.set_title(row_title[row_idx], prop={'size': 12})\n",
    "        \n",
    "        if row_idx == 1:\n",
    "            ax.set_ylabel('Cumulative probability', fontsize=14)\n",
    "                \n",
    "        if row_idx == 2:\n",
    "            neg_ticks = np.arange(0, bins.min(), -10)[::-1]\n",
    "            pos_ticks = np.arange(0, bins.max(), 10)\n",
    "\n",
    "            x_ticks = np.concatenate([neg_ticks, pos_ticks[1:]])\n",
    "            x_ticks = x_ticks.astype(np.int8)\n",
    "            ax.set_xticks(x_ticks)\n",
    "            ax.set_xticklabels(map(str, x_ticks), fontsize=12)\n",
    "            tt = ax.set_xlabel('Gene expression fold change (log2)', fontsize=14)\n",
    "            tt.set_position((1, 0))\n",
    "        else:\n",
    "            ax.set_xticks([])\n",
    "            \n",
    "        ax.set_xlim(bins.min(), bins.max())\n",
    "\n",
    "        y_ticks = [0., 0.5, 1]\n",
    "        ax.set_yticks(y_ticks)\n",
    "        ax.set_yticklabels(map(str, y_ticks), fontsize=12)\n",
    "        ax.set_ylim(-0.05, 1.05)\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "\n",
    "            \n",
    "        ax = axes[row_idx, 1]\n",
    "        \n",
    "        for c, hist in zip(data_colors, data):\n",
    "            if c == 'dimgrey':\n",
    "                style = 'dashed'\n",
    "                zorder = 0\n",
    "            else:\n",
    "                style = 'solid'\n",
    "                zorder = 2\n",
    "            assert np.isclose(hist.sum() * stepsize, 1., atol=1e-6), 'CDF not 1'\n",
    "            ax.plot(bins, np.cumsum(hist) * stepsize,\n",
    "                    lw=2, ls=style, c=c, zorder=zorder)\n",
    "            ax.axhline(0.75, 0, 1, c='lightgrey',\n",
    "                       ls='dashed', zorder=0, lw=1)\n",
    "            ax.axhline(0.5, 0, 1, c='lightgrey',\n",
    "                       ls='dashed', zorder=0, lw=1)\n",
    "            ax.axhline(0.25, 0, 1, c='lightgrey',\n",
    "                       ls='dashed', zorder=0, lw=1)\n",
    "            \n",
    "            ax.axvline(-3, 0, 1, c='black', alpha=0.5,\n",
    "                       ls='dashed', zorder=0, lw=1)\n",
    "            ax.axvline(3, 0, 1, c='black', alpha=0.5,\n",
    "                       ls='dashed', zorder=0, lw=1)\n",
    "            \n",
    "        ax.set_yticks([])\n",
    "        if row_idx == 2:\n",
    "            x_ticks = [-3, 0, 3]\n",
    "            ax.set_xticks(x_ticks)\n",
    "            ax.set_xticklabels(map(str, x_ticks), fontsize=12)\n",
    "        else:\n",
    "            ax.set_xticks([])\n",
    "        \n",
    "        ax.set_xlim(-5, 5)\n",
    "        #ax.set_ylim(-0.05, 1.05)\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "\n",
    "    return fig, []\n",
    "    \n",
    "    \n",
    "def create_boxplot(datapoints, comp, fkey):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    boxcolor = 'dimgrey'\n",
    "    medcolor = 'grey'\n",
    "    median_props = {'color': medcolor, 'linewidth': 2}\n",
    "    box_props = {'color': boxcolor, 'linewidth': 1}\n",
    "    whisker_props = {'color': boxcolor, 'linewidth': 1}\n",
    "    cap_props = {'color': boxcolor, 'linewidth': 1}\n",
    "        \n",
    "    \n",
    "    fig, axes = plt.subplots(figsize=fig_sizes['two_col']['square'],\n",
    "                             nrows=2, ncols=2, sharex=False, sharey=False,\n",
    "                             gridspec_kw={'height_ratios': [1, 2]})\n",
    "    data_lut = {(0, 0): ('diff', 'log2fc'), (0, 1): ('stable', 'log2fc'),\n",
    "                (1, 0): ('diff', 'min_tpm'), (1, 1): ('stable', 'min_tpm')}\n",
    "        \n",
    "    plt.subplots_adjust(hspace=0.05, wspace=0.2)\n",
    "    s1, s2 = comp.split('_vs_')\n",
    "    for row in range(2):\n",
    "        for col in range(2):\n",
    "            ax = axes[row, col]\n",
    "            gene_type, data_type = data_lut[(row, col)]\n",
    "            \n",
    "            subset = datapoints.loc[datapoints['genetype'] == gene_type, :].copy()\n",
    "            \n",
    "            if row == 0 and col == 0:\n",
    "                ax.set_title('{} v {}\\nDEG N={}'.format(s1, s2, subset.shape[0]),\n",
    "                             fontsize=14)\n",
    "                ax.set_ylabel('abs ( log2 fc )', fontsize=14)\n",
    "            elif row == 0 and col == 1:\n",
    "                ax.set_title('{} v {}\\nstable N={}'.format(s1, s2, subset.shape[0]),\n",
    "                             fontsize=14)\n",
    "                ax.axhline(2, 0, 1, lw=2, ls='dashed',\n",
    "                           color='lightgrey', zorder=0)\n",
    "            else:\n",
    "                pass\n",
    "            \n",
    "            if row == 1 and col == 0:\n",
    "                ax.set_ylabel('minimum expression (TPM)', fontsize=14)\n",
    "                xl_pos = ax.set_xlabel('HSP hits in gene bodies (B) or enhancers (E)', fontsize=14)\n",
    "                xl_pos.set_position((1, 0))\n",
    "                                    \n",
    "            boxes, boxlabels = prepare_plot_data(subset, data_type)\n",
    "                        \n",
    "            ax.set_xlim(0, len(boxes) + 1)\n",
    "                       \n",
    "            bb = ax.boxplot(boxes, sym=\"\", labels=None, widths=0.35,\n",
    "                            medianprops=median_props, boxprops=box_props,\n",
    "                            whiskerprops=whisker_props, capprops=cap_props)\n",
    "            max_y = 0\n",
    "            for cap in bb['caps']:\n",
    "                max_y = max(max_y, cap.get_ydata()[0])\n",
    "            if max_y < 12:\n",
    "                ax.set_ylim(-0.5, max_y // 2 * 2 + 2)\n",
    "                yticks = np.arange(0, np.ceil(max_y), 3, dtype=np.int8)\n",
    "            else:\n",
    "                ax.set_ylim(-0.5, max_y // 5 * 5 + 5)\n",
    "                yticks = np.arange(0, np.ceil(max_y), 5, dtype=np.int8)\n",
    "            \n",
    "            ax.set_yticks(yticks)\n",
    "            ax.set_yticklabels(map(str, yticks), fontsize=12)\n",
    "            \n",
    "            if row == 0:\n",
    "                ax.set_xticks([])\n",
    "            if row == 1:\n",
    "                ax.set_xticklabels(boxlabels, fontsize=12, rotation=90)\n",
    "            if col == 0:\n",
    "                ax.axvline(1, 0, 1, lw=25, ls='solid', color='red',\n",
    "                           alpha=0.1, zorder=0)\n",
    "            \n",
    "    return fig, []\n",
    "\n",
    "\n",
    "def dump_genetrail_lists():\n",
    "    cache_file = os.path.join(cache_dir, '{}_plot_gene-enh-hsp_tpm-bin.h5'.format(date))\n",
    "    if not os.path.isfile(cache_file):\n",
    "        raise RuntimeError('No cache file detected')\n",
    "    dump_folder = genetrail_out + '_miRNA'\n",
    "    os.makedirs(dump_folder, exist_ok=True)\n",
    "        \n",
    "    with pd.HDFStore(cache_file, 'r') as hdf:\n",
    "        cached_keys = list(hdf.keys())\n",
    "        for seg in segmentations:\n",
    "            if seg != 'cmm18':\n",
    "                continue\n",
    "            for scoring in scorings:\n",
    "                if scoring != 'emission':\n",
    "                    continue\n",
    "                load_keys = [k for k in cached_keys if seg in k and scoring in k]\n",
    "                \n",
    "                for k in load_keys:\n",
    "                    comp = k.split('/')[-1]\n",
    "                    if comp not in ['Ma_vs_Mo', 'He_vs_Mo']:\n",
    "                        continue\n",
    "                    print('Dumping GeneTrail lists for {}'.format(comp))\n",
    "                    data = hdf[k]\n",
    "                    \n",
    "                    base_file = '_'.join([seg, scoring, comp, '{}']) + '.tsv'\n",
    "                    deg = data.loc[data['genetype'] == 'diff', :].copy()\n",
    "                    \n",
    "                    select_miss = np.logical_and(deg['body_hsp'] == 0,\n",
    "                                                 deg['total_hits'] == 0)\n",
    "                    miss = deg.loc[select_miss, :]\n",
    "                    miss_file = os.path.join(dump_folder, base_file.format('deg-miss'))\n",
    "                    with open(miss_file, 'w') as dump:\n",
    "                        _ = dump.write('\\n'.join(miss.index.tolist()) + '\\n')\n",
    "                    \n",
    "                    select_hit = np.logical_or(deg['body_hsp'] > 0,\n",
    "                                               deg['total_hits'] > 0)\n",
    "                    hit = deg.loc[select_hit, :]                    \n",
    "                    hit_file = os.path.join(dump_folder, base_file.format('deg-hit'))\n",
    "                    with open(hit_file, 'w') as dump:\n",
    "                        _ = dump.write('\\n'.join(hit.index.tolist()) + '\\n')\n",
    "                        \n",
    "                    bmiss = deg.loc[deg['body_hsp'] == 0, :]\n",
    "                    bmiss_file = os.path.join(dump_folder, base_file.format('deg-body-miss'))\n",
    "                    with open(bmiss_file, 'w') as dump:\n",
    "                        _ = dump.write('\\n'.join(bmiss.index.tolist()) + '\\n')\n",
    "                    \n",
    "                    bhit = deg.loc[deg['body_hsp'] > 0, :]\n",
    "                    bhit_file = os.path.join(dump_folder, base_file.format('deg-body-hit'))\n",
    "                    with open(bhit_file, 'w') as dump:\n",
    "                        _ = dump.write('\\n'.join(bhit.index.tolist()) + '\\n')\n",
    "                                            \n",
    "                    stg = data.loc[data['genetype'] == 'stable', :].copy()\n",
    "                    stg_file = os.path.join(dump_folder, base_file.format('seg-any'))\n",
    "                    with open(stg_file, 'w') as dump:\n",
    "                        _ = dump.write('\\n'.join(stg.index.tolist()) + '\\n')\n",
    "                        \n",
    "                    stgmiss = stg.loc[stg['body_hsp'] == 0, :]\n",
    "                    stgmiss_file = os.path.join(dump_folder, base_file.format('seg-body-miss'))\n",
    "                    with open(stgmiss_file, 'w') as dump:\n",
    "                        _ = dump.write('\\n'.join(stgmiss.index.tolist()) + '\\n')\n",
    "\n",
    "    return\n",
    "\n",
    "\n",
    "def print_mirna_info(genes, mirna, cell):\n",
    "    \n",
    "    print('Cell ', cell)\n",
    "    select = np.logical_and(mirna['name'].isin(genes.index),\n",
    "                            mirna['celltype'] == cell)\n",
    "    sub = mirna.loc[select, ['norm_count', 'mirna_id']]\n",
    "    print('Unique miRNAs ', sub['mirna_id'].unique().size)\n",
    "    dist = np.percentile(sub['norm_count'], [25, 50, 75, 95, 99])\n",
    "    print('Count distribution ', dist)\n",
    "    return\n",
    "\n",
    "\n",
    "def make_mirna_scatter(fg, bg, mirna, s1, s2, fk='X'):\n",
    "    \n",
    "    fig, (fg_ax, bg_ax) = plt.subplots(figsize=fig_sizes['two_col']['half'],\n",
    "                                       nrows=1, ncols=2, sharex=False, sharey=False)\n",
    "    \n",
    "    fg_mirna = mirna.loc[mirna['name'].isin(fg['name']), ['name', 'mirna_id', 'norm_count', 'celltype']]\n",
    "    fg_ids = set(fg_mirna['mirna_id'].unique())\n",
    "    print('FG ', len(fg_ids))\n",
    "    fg_s1 = fg_mirna.loc[fg_mirna['celltype'] == s1, ['name', 'norm_count']]\n",
    "    fg_s1['norm_count'] = np.log10(fg_s1['norm_count'] + 1)\n",
    "    fg_s1 = fg_s1.merge(fg, on='name', how='outer')\n",
    "    fg_s1.fillna({'norm_count': 0}, inplace=True)\n",
    "        \n",
    "    fg_s2 = fg_mirna.loc[fg_mirna['celltype'] == s2, ['name', 'norm_count']]\n",
    "    fg_s2['norm_count'] = -1 * np.log10(fg_s2['norm_count'] + 1)\n",
    "    fg_s2 = fg_s2.merge(fg, on='name', how='outer')\n",
    "    fg_s2.fillna({'norm_count': 0}, inplace=True)\n",
    "    \n",
    "    fg_ax.set_xlim(min(fg_s1['log2fc'].min(), fg_s2['log2fc'].min() - 1),\n",
    "                   max(fg_s1['log2fc'].max(), fg_s2['log2fc'].max()) + 1)\n",
    "    fg_ax.set_ylim(fg_s2['norm_count'].min() - 0.5,\n",
    "                   fg_s1['norm_count'].max() + 0.5)\n",
    "    \n",
    "    fg_ax.scatter(fg_s1['log2fc'], fg_s1['norm_count'],\n",
    "                  c='dodgerblue', alpha=0.7,\n",
    "                  marker='.', s=20)\n",
    "    \n",
    "    fg_ax.scatter(fg_s2['log2fc'], fg_s2['norm_count'],\n",
    "                  c='orange', alpha=0.7,\n",
    "                  marker='.', s=20)\n",
    "    \n",
    "    \n",
    "    \n",
    "    bg_mirna = mirna.loc[mirna['name'].isin(bg['name']), ['name', 'mirna_id', 'norm_count', 'celltype']]\n",
    "    bg_ids = set(bg_mirna['mirna_id'].unique())\n",
    "    print('BG ', len(bg_ids))\n",
    "    bg_s1 = bg_mirna.loc[bg_mirna['celltype'] == s1, ['name', 'norm_count']]\n",
    "    bg_s1['norm_count'] = np.log10(bg_s1['norm_count'] + 1)\n",
    "    bg_s1 = bg_s1.merge(bg, on='name', how='outer')\n",
    "    bg_s1.fillna({'norm_count': 0}, inplace=True)\n",
    "        \n",
    "    bg_s2 = bg_mirna.loc[bg_mirna['celltype'] == s2, ['name', 'norm_count']]\n",
    "    bg_s2['norm_count'] = -1 * np.log10(bg_s2['norm_count'] + 1)\n",
    "    bg_s2 = bg_s2.merge(bg, on='name', how='outer')\n",
    "    bg_s2.fillna({'norm_count': 0}, inplace=True)\n",
    "    \n",
    "    bg_ax.set_xlim(min(bg_s1['log2fc'].min(), bg_s2['log2fc'].min() - 1),\n",
    "                   max(bg_s1['log2fc'].max(), bg_s2['log2fc'].max()) + 1)\n",
    "    bg_ax.set_ylim(bg_s2['norm_count'].min() - 0.5,\n",
    "                   bg_s1['norm_count'].max() + 0.5)\n",
    "    \n",
    "    bg_ax.scatter(bg_s1['log2fc'], bg_s1['norm_count'],\n",
    "                  c='dodgerblue', alpha=0.7,\n",
    "                  marker='.', s=20)\n",
    "    \n",
    "    bg_ax.scatter(bg_s2['log2fc'], bg_s2['norm_count'],\n",
    "                  c='orange', alpha=0.7,\n",
    "                  marker='.', s=20)\n",
    "    print(len(fg_ids.intersection(bg_ids)))\n",
    "    \n",
    "    \n",
    "    return fig, []\n",
    "\n",
    "\n",
    "def plot_mirna_changes():\n",
    "    cache_file = os.path.join(cache_dir, '{}_plot_gene-enh-hsp_tpm-bin.h5'.format(date))\n",
    "    if not os.path.isfile(cache_file):\n",
    "        raise RuntimeError('No cache file detected')\n",
    "        \n",
    "    mirna_cache = os.path.join(fhgfs_base, 'caching', 'notebooks', '20180711_mirna_exp_targets.h5')\n",
    "    if not os.path.isfile(mirna_cache):\n",
    "        raise RuntimeError('No miRNA annotation detected')\n",
    "    print('Cached data available...')\n",
    "    \n",
    "    with pd.HDFStore(cache_file, 'r') as hdf:\n",
    "        cached_keys = list(hdf.keys())\n",
    "        for seg in segmentations:\n",
    "            if seg != 'cmm18':\n",
    "                continue\n",
    "            for scoring in scorings:\n",
    "                if scoring != 'emission':\n",
    "                    continue\n",
    "                load_keys = [k for k in cached_keys if seg in k and scoring in k]\n",
    "                \n",
    "                for k in load_keys:\n",
    "                    comp = k.split('/')[-1]\n",
    "                    if comp not in ['Ma_vs_Mo']:\n",
    "                        continue\n",
    "                    s1, s2 = comp.split('_vs_')\n",
    "                    data = hdf[k]\n",
    "                    de_mirna_file = os.path.join(base_out, 'de_mirna.tsv')\n",
    "                    de_mirna = pd.read_csv(de_mirna_file, sep='\\t', header=None,\n",
    "                                           names=['mirna_id', 'log2fc'])\n",
    "#                     print(de_mirna.shape)\n",
    "#                     print('Loading miRNA')\n",
    "#                     with pd.HDFStore(mirna_cache, 'r') as cache:\n",
    "#                         mirna = cache['mirna_exp/' + comp]\n",
    "#                         mirna = mirna.loc[mirna['mirna_id'].isin(de_mirna['mirna_id']), :].copy()\n",
    "#                         mirna.drop_duplicates(['mirna_id', 'name'], inplace=True)\n",
    "                    ref_folder = os.path.join(fhgfs_base, 'references')\n",
    "                    trg_scan = 'TargetScan_v7.2_default_predictions.txt.zip'\n",
    "                    mirna = pd.read_csv(os.path.join(ref_folder, trg_scan), sep='\\t', header=0)\n",
    "                    mirna = mirna.loc[mirna['Species ID'] == 9606, :].copy()\n",
    "                    mirna['name'] = mirna['Gene ID'].str.extract('(?P<ENSID>\\w+)', expand=False)\n",
    "                    mirna['enst'] = mirna['Transcript ID'].str.extract('(?P<ENSID>\\w+)', expand=False)\n",
    "                    mirna['hit_start'] = mirna['UTR start']\n",
    "                    mirna['hit_end'] = mirna['UTR end']\n",
    "                    mirna['mirna_family'] = 'hsa-' + mirna['miR Family']\n",
    "                    mirna = mirna.loc[:, ['mirna_family', 'name', 'enst', 'hit_start', 'hit_end']]\n",
    "                    \n",
    "                    split_set = []\n",
    "                    de_mirna_ids = set(de_mirna['mirna_id'])\n",
    "                    for row in mirna.itertuples():\n",
    "                        members = row.mirna_family.split('/')\n",
    "                        for m in members:\n",
    "                            if m.startswith('hsa-miR'):\n",
    "                                keep = m\n",
    "                            else:\n",
    "                                keep = 'hsa-miR-' + row.mirna_family\n",
    "                            if keep in de_mirna_ids and row.name in data.index:\n",
    "                                split_set.append((keep, row.name, row.enst, row.hit_start, row.hit_end))\n",
    "                    mirna = pd.DataFrame(split_set, columns=['mirna_id', 'name', 'transcript',\n",
    "                                                             'hit_start', 'hit_end'])\n",
    "                    mirna.drop_duplicates(['mirna_id', 'transcript', 'hit_start'])\n",
    "                    print(col.Counter(mirna['mirna_id']).most_common(10))                                                     \n",
    "                    data['group'] = 'empty'\n",
    "                    data['name'] = data.index\n",
    "                    data.reset_index(drop=True, inplace=True)\n",
    "                                        \n",
    "                    group_a = np.logical_and(data['genetype'] == 'diff',\n",
    "                                             np.logical_or(data['body_hsp'] > 0,\n",
    "                                                           data['total_hits'] > 0))\n",
    "                    data.loc[group_a, 'group'] = 'deg_hsp-hit'\n",
    "                    \n",
    "                    group_b = np.logical_and(data['genetype'] == 'stable',\n",
    "                                             np.logical_or(data['body_hsp'] > 0,\n",
    "                                                           data['total_hits'] > 0))\n",
    "                    data.loc[group_b, 'group'] = 'stable_hsp-hit'\n",
    "                    \n",
    "                    group_c = np.logical_and(data['genetype'] == 'diff',\n",
    "                                             np.logical_and(data['body_hsp'] == 0,\n",
    "                                                            data['total_hits'] == 0))\n",
    "                    data.loc[group_c, 'group'] = 'deg_no-hit'\n",
    "                    \n",
    "                    group_d = np.logical_and(data['genetype'] == 'stable',\n",
    "                                             np.logical_and(data['body_hsp'] == 0,\n",
    "                                                            data['total_hits'] == 0))\n",
    "                    data.loc[group_d, 'group'] = 'stable_no-hit'\n",
    "                    \n",
    "                    src_group = col.Counter(data['group'])\n",
    "\n",
    "                    data = data.loc[data['name'].isin(mirna['name']), :].copy()\n",
    "                    group_count = col.Counter(data['group'])\n",
    "                    data = data.merge(mirna, on='name', how='outer')\n",
    "                    print(data.shape)\n",
    "                    hit_counts = col.Counter(data['group'])\n",
    "                                        \n",
    "                    for g, hits in hit_counts.items():\n",
    "                        gsize = group_count[g]\n",
    "                        print('Group ', g)\n",
    "                        print('Group size ', gsize, '({})'.format(src_group[g]))\n",
    "                        print('Target hits in group ', hits)\n",
    "                        print('Avg. target hits per gene ', np.round(hits / gsize, 2))\n",
    "                        print('====')\n",
    "                    \n",
    "                    \n",
    "#                     fig, exart = make_mirna_scatter(fg, bg, mirna, s1, s2, 'X')\n",
    "    return\n",
    "\n",
    "\n",
    "if run_dump_genetrail_lists:\n",
    "    #dump_genetrail_lists()\n",
    "    plot_mirna_changes()\n",
    "    \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
