{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dumping  /TL/deep/fhgfs/projects/pebert/thesis/projects/statediff/chromdiff/exec_scripts/chromdiff_HG_vs_He.sh\n",
      "Dumping  /TL/deep/fhgfs/projects/pebert/thesis/projects/statediff/chromdiff/exec_scripts/chromdiff_HG_vs_Ma.sh\n",
      "Dumping  /TL/deep/fhgfs/projects/pebert/thesis/projects/statediff/chromdiff/exec_scripts/chromdiff_HG_vs_Mo.sh\n",
      "Dumping  /TL/deep/fhgfs/projects/pebert/thesis/projects/statediff/chromdiff/exec_scripts/chromdiff_He_vs_Ma.sh\n",
      "Dumping  /TL/deep/fhgfs/projects/pebert/thesis/projects/statediff/chromdiff/exec_scripts/chromdiff_He_vs_Mo.sh\n",
      "Dumping  /TL/deep/fhgfs/projects/pebert/thesis/projects/statediff/chromdiff/exec_scripts/chromdiff_Ma_vs_Mo.sh\n",
      "Dumping  /TL/deep/fhgfs/projects/pebert/thesis/projects/statediff/chromdiff/exec_scripts/chromdiff_Liver_vs_Blood.sh\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import os as os\n",
    "import re as re\n",
    "import csv as csv\n",
    "import gzip as gzip\n",
    "import io as io\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import shutil as sh\n",
    "\n",
    "fhgfs_base = '/TL/deep/fhgfs/projects/pebert/thesis/projects/statediff'\n",
    "\n",
    "chromhmm_segment = os.path.join(fhgfs_base, 'chromhmm/deep/segmentation')\n",
    "epigenome_colors = {'LiHG_Ct': '#a6cee3'.upper(),\n",
    "                    'LiHe_Ct': '#1f78b4'.upper(),\n",
    "                    'BlMo_Ct': '#b2df8a'.upper(),\n",
    "                    'BlMa_Ct': '#33a02c'.upper()}\n",
    "output_dir = os.path.join(fhgfs_base, 'chromdiff/input_files')\n",
    "\n",
    "def step0_adapt_state_names():\n",
    "    \n",
    "    segfiles = sorted(os.listdir(chromhmm_segment))\n",
    "    records = []\n",
    "    for sf in segfiles:\n",
    "        if 'LiHe' in sf:\n",
    "            if not ('Hf02' in sf or 'Hf03' in sf):\n",
    "                continue\n",
    "        if sf.startswith('52_'):\n",
    "            continue\n",
    "        out_buffer = io.StringIO()\n",
    "        new_file = os.path.join(output_dir, sf)\n",
    "        if os.path.isfile(new_file):\n",
    "            continue\n",
    "        with open(os.path.join(chromhmm_segment, sf), 'r') as bedfile:\n",
    "            for line in bedfile:\n",
    "                if line:\n",
    "                    parts = line.strip().split()\n",
    "                    try:\n",
    "                        new_line = '\\t'.join(parts[:3] + [parts[3].strip('E')])\n",
    "                    except IndexError:\n",
    "                        print(line)\n",
    "                        raise\n",
    "                    out_buffer.write(new_line + '\\n')\n",
    "        with open(new_file, 'w') as dump:\n",
    "            _ = dump.write(out_buffer.getvalue())\n",
    "    return\n",
    "    \n",
    "\n",
    "def step1_epigenome_annotation():\n",
    "    metadata_file = os.path.join(output_dir, 'sample_annotation.tsv')\n",
    "    metadata_header = ['ID', 'filepath', 'color', 'name', 'celltype', 'tissue']\n",
    "    #metadata_header = ['ID', 'filepath', 'color', 'name', 'celltype']\n",
    "    \n",
    "    tissue_map = {'Li': 'Liver', 'Bl': 'Blood'}\n",
    "    \n",
    "    segfiles = sorted(os.listdir(output_dir))\n",
    "    records = []\n",
    "    for sf in segfiles:\n",
    "        if not sf.endswith('segments.bed'):\n",
    "            continue\n",
    "        infos = {'filepath': os.path.join(output_dir, sf)}\n",
    "        parts = sf.split('_')\n",
    "        eid = '_'.join(parts[1:3])\n",
    "        try:\n",
    "            color = epigenome_colors['_'.join(parts[2:4])]\n",
    "        except KeyError:\n",
    "            continue\n",
    "        celltype = parts[2][2:]\n",
    "        infos['celltype'] = celltype\n",
    "        infos['tissue'] = tissue_map[parts[2][:2]]\n",
    "        infos['ID'] = eid\n",
    "        infos['color'] = color\n",
    "        infos['name'] = '_'.join(parts[1:4])\n",
    "        records.append(infos)\n",
    "    with open(metadata_file, 'w', newline='') as table:\n",
    "        writer = csv.DictWriter(table, metadata_header, delimiter='\\t')\n",
    "        writer.writeheader()\n",
    "        writer.writerows(records)\n",
    "    return metadata_file\n",
    "\n",
    "def step2_gene_annotation():\n",
    "    \"\"\"\n",
    "    Restrict to protein coding\n",
    "    genes and keep only chr1-22, X\n",
    "    \"\"\"\n",
    "    header = ['#chrom', 'start', 'end', 'strand', 'gencode_gene_id', 'genesymbols']\n",
    "    gencode_file = os.path.join(output_dir, 'gencode_v21_pc.bed')\n",
    "    \n",
    "    gencode_cache = os.path.join(output_dir, 'gencode_v21_pc.h5')\n",
    "    if os.path.isfile(gencode_cache):\n",
    "        with pd.HDFStore(gencode_cache, 'r') as hdf:\n",
    "            df = hdf['genes']\n",
    "        return gencode_file, df\n",
    "    \n",
    "    gencode_pc_genes = os.path.join(fhgfs_base, 'references', 'gencode.v21.pc_transcripts.fa.gz')\n",
    "    gencode_genes = os.path.join(fhgfs_base, 'references', 'gencode.v21.annotation.gtf.gz')\n",
    "    transcripts = set()\n",
    "    genes = set()\n",
    "    with gzip.open(gencode_pc_genes, 'rt') as fasta:\n",
    "        for line in fasta:\n",
    "            if line.startswith('>'):\n",
    "                parts = line.strip('>').split('|')\n",
    "                transcripts.add(parts[0].strip())\n",
    "                genes.add(parts[1].strip())\n",
    "    keep_chroms = ['chr' + str(x) for x in range(1, 23, 1)]\n",
    "    keep_chroms.append('chrX')\n",
    "    \n",
    "    get_name = re.compile('gene_name\\s\\\"(?P<SYMBOL>[A-Z0-9\\.a-z\\-_\\:]+)\\\"')\n",
    "    get_geneid = re.compile('gene_id\\s\\\"(?P<GENEID>ENS[A-Z0-9\\.a-z]+)\\\"')\n",
    "    \n",
    "    locations = []\n",
    "    with gzip.open(gencode_genes, 'rt') as gtf:\n",
    "        for line in gtf:\n",
    "            if line.startswith('chr'):\n",
    "                parts = line.split('\\t')\n",
    "                c, _, entity, s, e, _, strand, _, attr = parts\n",
    "                if c in keep_chroms and entity == 'gene':\n",
    "                    mobj = get_geneid.search(attr)\n",
    "                    gene_id = mobj.group('GENEID')\n",
    "                    mobj = get_name.search(attr)\n",
    "                    assert mobj is not None, 'Could not find name: {}'.format(attr)\n",
    "                    gene_symbol = mobj.group('SYMBOL')\n",
    "                    if gene_id in genes:\n",
    "                        locations.append((c, s, e, strand, gene_id.strip(), gene_symbol.strip()))\n",
    "    df = pd.DataFrame(locations, columns=['chrom', 'start', 'end', 'strand', 'name', 'symbol'])\n",
    "    with open(gencode_file, 'w') as dump:\n",
    "        _ = dump.write('#')\n",
    "        df.to_csv(dump, sep='\\t', header=True, index=False,\n",
    "                  index_label=False, line_terminator='\\n')\n",
    "    with pd.HDFStore(gencode_cache, 'w') as hdf:\n",
    "        hdf.put('genes', df, format='table')\n",
    "    return gencode_file, df\n",
    "                \n",
    "def step3_state_annotation():\n",
    "    annotation_src = '/home/pebert/work/code/mpggit/statediff/annotation/chromhmm_18/chromhmm_18_states_info.tsv'\n",
    "    fname = os.path.basename(annotation_src)\n",
    "    annotation_dest = os.path.join(output_dir, fname)\n",
    "    sh.copyfile(annotation_src, annotation_dest)\n",
    "    return annotation_dest\n",
    "    \n",
    "def step4_expression_table(genes):\n",
    "    exptab_file = os.path.join(output_dir, 'pcgene_exp_tpm.tsv')\n",
    "    exp_root = os.path.join(fhgfs_base, 'salmon/deep/quant')\n",
    "    exp_table = []\n",
    "    for root, dirs, datafiles in os.walk(exp_root):\n",
    "        if root.endswith('_Ct'):\n",
    "            quant = [f for f in datafiles if f == 'quant.genes.sf']\n",
    "            quant_file = os.path.join(root, quant[0])\n",
    "            eid = os.path.split(root)[-1]\n",
    "            eid = eid[3:12]\n",
    "            df = pd.read_csv(quant_file, usecols=['Name', 'TPM'], sep='\\t', index_col=0)\n",
    "            df.columns = [eid]\n",
    "            exp_table.append(df)\n",
    "    exp_table = pd.concat(exp_table, axis=1, ignore_index=False)\n",
    "    exp_table = exp_table.loc[exp_table.index.isin(genes['name']), :].copy()\n",
    "    exp_table.index.name = 'gene_id'\n",
    "    exp_table.to_csv(exptab_file, sep='\\t', header=True, index=True)\n",
    "    return exptab_file\n",
    "            \n",
    "def step5_covariate_matrix(smp_ann):\n",
    "    covmat_file = os.path.join(output_dir, 'covmat.tsv')\n",
    "    md = pd.read_csv(smp_ann, sep='\\t')\n",
    "    md = md[['ID', 'celltype', 'tissue']]\n",
    "#     md = md[['ID', 'celltype']]\n",
    "    md.index = md['ID']\n",
    "    md.drop(['ID'], axis=1, inplace=True)\n",
    "    \n",
    "    num_cols = md['celltype'].unique().size + md['tissue'].unique().size\n",
    "    col_names = md['celltype'].unique().tolist() + md['tissue'].unique().tolist()\n",
    "    num_rows = md.shape[0]\n",
    "    \n",
    "    cov_mat = pd.DataFrame(np.zeros((num_rows, num_cols), dtype=np.int8),\n",
    "                           index=md.index, columns=col_names)\n",
    "    for row in md.itertuples():\n",
    "        cov_mat.loc[row.Index, row.celltype] = 1\n",
    "        cov_mat.loc[row.Index, row.tissue] = 1\n",
    "    cov_mat.to_csv(covmat_file, sep='\\t', index=True, index_label=False, header=True)\n",
    "    return covmat_file\n",
    "\n",
    "def step6_map_vars_cov():\n",
    "    map_vars_file = os.path.join(output_dir, 'map_vars_cov.tsv')\n",
    "    with open(map_vars_file, 'w') as table:\n",
    "        _ = table.write('\\t'.join(['HG' , 'HE', 'MO', 'MA', 'LIVER', 'BLOOD']) + '\\n')\n",
    "        _ = table.write('CELLTYPE' + '\\t'.join(['', 'TRUE', 'TRUE', 'TRUE', 'TRUE', 'FALSE', 'FALSE']) + '\\n')\n",
    "        _ = table.write('TISSUE' + '\\t'.join(['', 'FALSE', 'FALSE', 'FALSE', 'FALSE', 'TRUE', 'TRUE']) + '\\n')\n",
    "#         _ = table.write('\\t'.join(['HG' , 'HE', 'MO', 'MA']) + '\\n')\n",
    "#         _ = table.write('CELLTYPE' + '\\t'.join(['', 'TRUE', 'TRUE', 'TRUE', 'TRUE']) + '\\n')\n",
    "    return map_vars_file\n",
    "    \n",
    "def create_run_scripts(annotations):\n",
    "    chromdiff_script = '/home/pebert/work/code/github/ChromDiff/notes_v2.sh'\n",
    "    replace_lines = {'metadatafile': 'metadatafile=\"{smp_ann_fpath}\"',\n",
    "                     'genefile': 'genefile=\"{genes_fpath}\"',\n",
    "                     'generegions_label': 'generegions_label=\"gencode_v21\"',\n",
    "                     'statecalls_label': 'statecalls_label=\"deep\"',\n",
    "                     'states_info': 'states_info=\"{state_ann_fpath}\"',\n",
    "                     'covariate_mat_file': 'covariate_mat_file=\"{cov_mat_fpath}\"',\n",
    "                     'map_covariates_file': 'map_covariates_file=\"{map_vars_fpath}\"',\n",
    "                     'expfile': 'expfile=\"{exptable_fpath}\"',\n",
    "                     'state_annotations_file': 'state_annotations_file=\"{state_ann_fpath}\"',\n",
    "                     'property': 'property=\"{property}\"',\n",
    "                     'a_option': 'a_option=\"{a_group}\"',\n",
    "                     'b_option': 'b_option=\"{b_group}\"',\n",
    "                     'curr_label': \"curr_label=${{statecalls_label}}_${{generegions_label}}\"}\n",
    "    comp_groups = [('celltype', 'HG', 'He'), ('celltype','HG', 'Ma'),\n",
    "                   ('celltype','HG', 'Mo'), ('celltype', 'He', 'Ma'),\n",
    "                   ('celltype', 'He', 'Mo'), ('celltype', 'Ma', 'Mo'),\n",
    "                   ('tissue', 'Liver', 'Blood')]\n",
    "    \n",
    "    buffer = io.StringIO()\n",
    "    with open(chromdiff_script, 'r') as bash:\n",
    "        for line in bash:\n",
    "            if line.startswith('## then look at'):\n",
    "                continue\n",
    "            elif line.startswith('#'):\n",
    "                buffer.write(line)\n",
    "            else:\n",
    "                rep = replace_lines.get(line.split('=')[0], line.strip())\n",
    "                buffer.write(rep + '\\n')\n",
    "    script_out = os.path.join(fhgfs_base, 'chromdiff/exec_scripts')\n",
    "    script_out2 = '/home/pebert/work/code/github/ChromDiff'\n",
    "    for (p, a, b) in comp_groups:\n",
    "        tmp = annotations\n",
    "        tmp.update({'a_group': a,\n",
    "                    'b_group': b,\n",
    "                    'property': p})\n",
    "        script = buffer.getvalue()\n",
    "        script = script.format(**tmp)\n",
    "        script_name = 'chromdiff_{}_vs_{}.sh'.format(a, b)\n",
    "        script_path = os.path.join(script_out, script_name)\n",
    "        print('Dumping ', script_path)\n",
    "        with open(script_path, 'w') as dump:\n",
    "            _ = dump.write(script)\n",
    "        script_path = os.path.join(script_out2, script_name)\n",
    "        with open(script_path, 'w') as dump:\n",
    "            _ = dump.write(script)\n",
    "    return              \n",
    "    \n",
    "step0_adapt_state_names()\n",
    "ann_files = dict()\n",
    "sample_file = step1_epigenome_annotation()\n",
    "ann_files['smp_ann_fpath'] = sample_file\n",
    "gencode_file, genes = step2_gene_annotation()\n",
    "ann_files['genes_fpath'] = gencode_file\n",
    "state_file = step3_state_annotation()\n",
    "ann_files['state_ann_fpath'] = state_file\n",
    "exptab_file = step4_expression_table(genes)\n",
    "ann_files['exptable_fpath'] = exptab_file\n",
    "covmat_file = step5_covariate_matrix(sample_file)\n",
    "ann_files['cov_mat_fpath'] = covmat_file\n",
    "mapvars_file = step6_map_vars_cov()\n",
    "ann_files['map_vars_fpath'] = mapvars_file\n",
    "create_run_scripts(ann_files)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
